kythe/rust/indexer/src/providers.rs

```diff
@@ -56,12 +58,34 @@ impl KzipFileProvider {
     ///
     /// # Errors
     ///
-    /// If the file is not a valid zip archive, a
+    /// If an error occurs while reading the kzip, a
     /// [KzipFileError][KytheError::KzipFileError] will be returned.
     pub fn new(file: File) -> Result<Self, KytheError> {
         let reader = BufReader::new(file);
-        let zip_archive = ZipArchive::new(reader)?;
-        Ok(Self { zip_archive })
+        let mut zip_archive = ZipArchive::new(reader)?;
+
+        // Get the name of the root folder of the kzip. This should be almost always be
+        // "root" by the kzip spec doesn't guarantee it.
+        let root_name: String;
+        // Extra scrope is needed because we are borrowing zip_archive for file. File
+        // needs to get dropped so release the zip_archive borrow before we move
+        // it into a struct.
+        {
+            let file = zip_archive.by_index(0)?;
+            let mut path = Path::new(file.name());
+            while let Some(p) = path.parent() {
+                // Last parent will be blank path so break when p is empty. path will contain
+                // root
+                if p == Path::new("") {
+                    break;
+                }
+                path = p;
+            }
+            root_name = path.to_str().unwrap().into();
+        }
+        // Safe to unwrap because kzip read would have failed if the internal paths
```

Comment:
This comment looks like should be above line 84.

Reply:
Ah you're right